Src/GoogleApis.Auth/OAuth2/ClientCredentialParameters.cs

```diff
@@ -0,0 +1,105 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+using Google.Apis.Json;
+using Google.Apis.Util;
+
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>Indicates the type of Client Credential.</summary>
+    public enum ClientCredentialType
+    {
+        /// <summary>UserCredential is created by the gcloud sdk tool when the user runs <a href="https://cloud.google.com/sdk/gcloud/reference/auth/login">gcloud auth login</a>.</summary>
+        UserCredential = 1,
+        /// <summary>ServiceAccountCredential is downloaded by the user from <a href="https://console.developers.google.com">Google Developers Console</a>.</summary>
+        ServiceAccountCredential = 2
+    }
+
+    /// <summary>Container for the union of parameters/properties from all credential types.</summary>
+    public sealed class ClientCredentialParameters
+    {
+        /// <summary>Indicates whether the credentials is of type ServiceAccountCredential or UserCredential.</summary>
+        [Newtonsoft.Json.JsonProperty("type")]
+        private string Type { get; set; }
+
+        /// <summary>Client Email associated with ServiceAccountCredential obtained from <a href="https://console.developers.google.com">Google Developers Console</a></summary>
+        [Newtonsoft.Json.JsonProperty("client_email")]
+        public string ClientEmail { get; set; }
+
+        /// <summary>Cliend Id associated with UserCredential created by <a href="https://cloud.google.com/sdk/gcloud/reference/auth/login">gcloud auth login</a>.</summary>
+        [Newtonsoft.Json.JsonProperty("client_id")]
+        private string ClientId { get; set; }
+
+        /// <summary>Client Secret associated with UserCredential created by <a href="https://cloud.google.com/sdk/gcloud/reference/auth/login">gcloud auth login</a>.</summary>
+        [Newtonsoft.Json.JsonProperty("client_secret")]
+        private string ClientSecret { get; set; }
+
+        /// <summary>Private Key associated with ServiceAccountCredential obtained from <a href="https://console.developers.google.com">Google Developers Console</a>.</summary>
+        [Newtonsoft.Json.JsonProperty("private_key")]
+        public string Pkcs8PrivateKey { get; set; }
+
+        /// <summary>Refresh Token associated with UserCredential created by <a href="https://cloud.google.com/sdk/gcloud/reference/auth/login">gcloud auth login</a>.</summary>
+        [Newtonsoft.Json.JsonProperty("refresh_token")]
+        public string RefreshToken { get; set; }
+
+        /// <summary>Gets the client secrets which contains the client identifier and client secret. </summary>
+        public ClientCredentialType CredentialType
+        {
+            get
+            {
+                Utilities.ThrowIfNullOrEmpty(Type, "Type");
+
+                switch (Type)
+                {
+                    case "authorized_user":
```

Comment:
I've restructured the code as below to avoid the need for constants and also to make a little bit cleaner.

public enum ClientCredentialType
    {
        /// <summary>UserCredential is created by the gcloud sdk tool when the user runs <a href="https://cloud.google.com/sdk/gcloud/reference/auth/login">gcloud auth login</a>.</summary>
        [StringValue("user_credential")]
        UserCredential = 1,
        
        /// <summary>ServiceAccountCredential is downloaded by the user from <a href="https://console.developers.google.com">Google Developers Console</a>.</summary>
        [StringValue("service_account")]
        ServiceAccountCredential = 2
    }


return (ClientCredentialType)Enum.Parse(typeof(ClientCredentialType), Type, true);